### PR TITLE
Constant source node DOM

### PR DESCRIPTION
**Pull request by josephhutch**
This is the DOM implementation for ConstantSourceNode.  Additional modification is likely, but I wanted to get the review process started for guidance.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23152)

### CODE REVIEW

File: components/script/dom/constantsourcenode.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            1, /* outputs */
+        )?;
+        let node_id = source_node.node().node_id();
+        let offset = AudioParam::new(
+            window,
+            context,
+            node_id,
+            ParamType::Offset,
+            AutomationRate::A_rate,
+            1.,
```
Comment (Manishearth) [inline components/script/dom/constantsourcenode.rs]: You should use `options.offset` here.

**Review by Manishearth — APPROVED** [review]
This looks great so far! You'll want to uncomment the method on `BaseAudioContext` (and uncomment the corresponding line from `BaseAudioContext.webidl`. Once that's done, be sure to run `./mach test-wpt tests/wpt/web-platform-tests/webaudio/the-audio-api/the-constantsourcenode-interface` to see if tests pass.

(You can update test expectations following the instructions [here](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations), be sure to only run tests for the webaudio stuff so it doesn't take forever)

**Comment by ferjm** [comment]
Thanks for the PR @josephhutch

So far, so good :)

CI is complaining about [a couple of small things](https://travis-ci.com/servo/servo/jobs/189472276#L464) that should be resolved by running `./mach fmt` and uncommenting and finishing the implementation of `CreateConstantSource` (including a comment with a link to the spec).

There are some [Web Platform Tests](https://github.com/web-platform-tests/wpt#the-web-platform-tests-project-) [here](https://github.com/servo/servo/tree/master/tests/wpt/web-platform-tests/webaudio/the-audio-api/the-constantsourcenode-interface) that you can use to test the implementation. You can run them with `./mach test-wpt tests/wpt/web-platform-tests/webaudio/the-audio-api/the-constantsourcenode-interface --log-raw /tmp/servo.log`. There will likely be changes between the existing results and the results after applying your patch, so you will likely need to [update the test expectations](https://github.com/servo/servo/tree/master/tests/wpt#updating-test-expectations) before merging this PR: `./mach update-wpt /tmp/servo.log`.

**Comment by ferjm** [comment]
Woops, I didn't see @Manishearth's review :D

File: components/script/dom/constantsourcenode.rs (hunk 1 above)
Comment (hustlerman) [reply components/script/dom/constantsourcenode.rs]: Should this replace ParamType::Offset? Tried that before running the mach test command and the tests timed out. The tests ran with ParamType::Offset, though 3 out of 5 failed.

**Comment by hustlerman** [comment]
Hi, I applied the changes above, but had a couple of questions:
- I tried replacing 1. with options.offset as @Manishearth suggested, but I was getting a type error on build. Since our group was basing our code on OscillatorNode (which used magic numbers to indicate the defaults), I left this in until I know how to fix the type issue.
- I'm not entirely sure what tests I should expect to pass with these changes. Looks like only 1 out of the 5 succeeded with no errors, but that might also be resolved by the above point. The current test expectations/results were updated with the latest push.

**Comment by jdm** [comment]
@hustlerman What was the type error you received?

**Comment by hustlerman** [comment]
[chrisaben@localhost servo]$ ./mach build -d
   Compiling script v0.0.1 (/home/chrisaben/Documents/csc517/servo/components/script)
error[E0308]: mismatched types
  --> components/script/dom/constantsourcenode.rs:57:13
   |
57 |             options.offset,
   |             ^^^^^^^^^^^^^^ expected f32, found struct `dom::bindings::num::Finite`
   |
   = note: expected type `f32`
              found type `dom::bindings::num::Finite<f32>`

error: aborting due to previous error

**Comment by hustlerman** [comment]
My guess is that this is specified in ConstantSourceNode.webidl in line 10?
```
dictionary ConstantSourceOptions: AudioNodeOptions {
  float offset = 1;
};
```
However, it's specified similarly in other Nodes that use the options.parameter default. Taking another look into it.

**Comment by hustlerman** [comment]
Best that I can tell is that all of the Node Option classes are generated by the system since they fall under the codegen crate? I'm not entirely sure, but just let me know where I need to look.

**Comment by jdm** [comment]
You can find how to extract the float value from a Finite type from our docs: https://doc.servo.org

**Comment by hustlerman** [comment]
Alright, I tried both options.offset.into() and f32::from(options.offset) to no avail. 

Compiling script v0.0.1 (/home/chrisaben/Documents/csc517/servo/components/script)
error[E0277]: the trait bound `f32: std::convert::From<dom::bindings::num::Finite<f32>>` is not satisfied
  --> components/script/dom/constantsourcenode.rs:57:28
   |
57 |             options.offset.into(),
   |                            ^^^^ the trait `std::convert::From<dom::bindings::num::Finite<f32>>` is not implemented for `f32`
   |
   = help: the following implementations were found:
             <f32 as std::convert::From<i16>>
             <f32 as std::convert::From<i8>>
             <f32 as std::convert::From<style::values::computed::MozScriptSizeMultiplier>>
             <f32 as std::convert::From<style::values::generics::GreaterThanOrEqualToOne<f32>>>
           and 4 others
   = note: required because of the requirements on the impl of `std::convert::Into<f32>` for `dom::bindings::num::Finite<f32>`

error: aborting due to previous error

Compiling script v0.0.1 (/home/chrisaben/Documents/csc517/servo/components/script)
error[E0277]: the trait bound `f32: std::convert::From<dom::bindings::num::Finite<f32>>` is not satisfied
  --> components/script/dom/constantsourcenode.rs:57:13
   |
57 |             f32::from(options.offset),
   |             ^^^^^^^^^ the trait `std::convert::From<dom::bindings::num::Finite<f32>>` is not implemented for `f32`
   |
   = help: the following implementations were found:
             <f32 as std::convert::From<i16>>
             <f32 as std::convert::From<i8>>
             <f32 as std::convert::From<style::values::computed::MozScriptSizeMultiplier>>
             <f32 as std::convert::From<style::values::generics::GreaterThanOrEqualToOne<f32>>>
           and 4 others
   = note: required by `std::convert::From::from`

error: aborting due to previous error

I'm new to Rust, but these seemed like the most reasonable calls from the Finite Type reference page: https://doc.servo.org/script/dom/bindings/num/struct.Finite.html

**Comment by jdm** [comment]
You can use `*finite_value` to extract the wrapped f32 or f64 value: https://github.com/servo/servo/blob/328244684a8b1f98d515d5f40655b9cf7347affd/components/script/dom/bindings/num.rs#L40-L43

**Comment by jdm** [comment]
@hustlerman Are you planning to finish this work?

**Comment by jdm** [comment]
Closing due to lack of response.

**Pull request closed without merging**